### PR TITLE
Remove duplicated validation in branch and link package step

### PR DESCRIPTION
**Pull request by krauselukas**
The super class of the steps already takes care of validating
the presence of the source package and source project names through
the REQUIRED_KEYS constant.

### CODE REVIEW

File: src/api/app/models/workflow/step/branch_package_step.rb [hunk 1]
```diff
@@ -1,20 +1,13 @@
 class Workflow::Step::BranchPackageStep < ::Workflow::Step
   REQUIRED_KEYS = [:source_project, :source_package].freeze
 
-  validates :source_project_name, presence: true
-  validates :source_package_name, presence: true
```
Comment (dmarcoux) [inline src/api/app/models/workflow/step/branch_package_step.rb:5]: Those validations need to stay.

**Review by dmarcoux — CHANGES_REQUESTED** [review]

**Review by dmarcoux — DISMISSED** [review]

**Comment by dmarcoux** [comment]
We could also remove this:

https://github.com/openSUSE/open-build-service/blob/b74e17fbc4e7ca68f1a1be317bb71f6112ec4b4e/src/api/app/models/workflow/step/configure_repositories.rb#L5

File: src/api/app/models/workflow/step.rb [hunk 2]
```diff
@@ -37,16 +37,26 @@ def create_or_update_subscriptions(package, workflow_filters)
 
   protected
 
-  def validate_step_instructions
+  def validate_step_keys
```
Comment (dmarcoux) [inline src/api/app/models/workflow/step.rb:40]: I would keep the validations in a single method to not loop twice over the required keys. It's not much overheard, but the code could be like this:

```ruby
  def validate_step_instructions
    self.class::REQUIRED_KEYS.each do |required_key|
      unless step_instructions.key?(required_key)
        errors.add(:base, "The '#{required_key}' key is missing")
        next
      end

      errors.add(:base, "The '#{required_key}' key must provide a value") if step_instructions[required_key].blank?
    end
  end
```

File: src/api/app/models/workflow/step.rb [hunk 3]
```diff
@@ -37,16 +37,26 @@ def create_or_update_subscriptions(package, workflow_filters)
...
     self.class::REQUIRED_KEYS.each do |required_key|
       errors.add(:base, "The '#{required_key}' key is missing") unless step_instructions.key?(required_key)
     end
   end
 
+  def validate_step_values
+    self.class::REQUIRED_KEYS.each do |required_key|
+      errors.add(:base, "The '#{required_key}' key must provide a value") if step_instructions[required_key].blank?
```
Comment (dmarcoux) [inline src/api/app/models/workflow/step.rb:48]: If `required_key` isn't present in `step_instructions`, `step_instructions[required_key].blank?` will cause the following exception:

```
NoMethodError (undefined method `blank?' for nil:NilClass)
```

File: src/api/app/models/workflow/step.rb (hunk 2 above)
Comment (krauselukas) [reply src/api/app/models/workflow/step.rb]: :+1: done

**Comment by dmarcoux** [comment]
@krauselukas I pushed to fix the failing spec. The test suite should be passing now.

**Review by danidoni — APPROVED** [review]

**Comment by dmarcoux** [comment]
The CI check `coverage/coveralls` seems to be stuck... I'm merging

**Comment by krauselukas** [comment]
@dmarcoux merci

**Pull request merged**